docs/tradeoffs.md

```diff
@@ -0,0 +1,136 @@
+# Design Trade-offs
+
+This will be looking at CLI parser design trade offs from the lens of comparing
+[bpaf](https://docs.rs/bpaf) and [clap](https://docs.rs/clap).
+
+For anyone asking the question "which should I use?", the short answer would be
+- `bpaf` would work well for simple cases like example bins for a library
+  while giving the best build times and a reasonable to understand code with
+  the `derive` API
+- All other cases the answer is "it depends" as the two designs offer different
+  trade offs.   If you want to just choose one with little research that will
+  cover the most use cases, that will most likely be `clap`.
+
+## Static vs Dynamic Typing
+
+`bpaf`'s combinator API uses generics and macros to build a statically-typed
+parser, pushing development errors to compile time.  This minimizes binary
+size and runtime, only paying for what you use.
+
+The combinator approach tends towards yoda-speak though with [careful
+structuring by breaking down arguments into
+functions](https://github.com/pacak/bpaf/blob/aa6992931bbfbdca6390c87f4a76898f8db0ae47/examples/top_to_bottom.rs),
+a more straightforward ordering can be accomplished.
+
+`clap`'s builder API stores parsed values in a Map-like container where the
+keys are the argument IDs and the values are effectively `Box<dyn Any>`.
+- This allows dynamically created CLIs, like with
+  [clap_serde](https://docs.rs/clap_serde) or [conditionally-present
+  flags](https://github.com/sharkdp/bat/blob/6680f65e4b25b0f18c455f7a4639a96e97519dc5/src/bin/bat/clap_app.rs#L556)
+- Callers can dynamically iterate over the arguments, like for layered configs.
+- `clap` can have ready-to-use, arbitrarily deifned defaults and validation
+  conditioned on argument values (e.g. `default_value_if`).  See the section on
+  Validation for more nuance on the "arbitrarily" part
+- `clap` tries to help push as many errors to early test time with minimal
+  effort with the `cmd.debug_assert()` function.  Value lookups still requires
```

Comment:
This is mostly trying to contrast with bpaf's compile-time reporting.  "Its not as early but it can still be relatively early" is the intent.  Later I talk about how both provide asserting mechanisms

Reply:
I see, maybe should be reworded to mention `bpaf` here, to me it reads as a separate feature. But again, I'm jetlagged :)

Reply:
Did a direct push contrasting it with compile time errors.